### PR TITLE
API-543: update product model with media file via API

### PR DESCRIPTION
**Pull request by willy-ahva**
Be able to send update product model with media files.

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added Behats                      | -
| Added integration tests           | OK
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: CHANGELOG-2.0.md [hunk 1]
```diff
@@ -12,17 +12,25 @@
...
+
+- API-543: update product model with media file via API
+
 # 2.0.10 (2017-12-22)
 
 ## Bug fixes
 
 - PIM-7032: Fix close button when clicking on "Compare/Translate" option on the product edit form
 - PIM-7063: Add validation for AttributeGroup - cannot remove AttributeGroup containing attributes and cannot remove AttributGroup "other"
+- PIM-7062: Allow to delete attributes contained in a family variant
```
Comment (momoss) [inline CHANGELOG-2.0.md]: Is it normal?

File: src/Pim/Bundle/ApiBundle/Resources/config/controllers.yml [hunk 2]
```diff
@@ -201,8 +201,11 @@ services:
             - '@akeneo_file_storage.file_storage.filesystem_provider'
             - '@akeneo_file_storage.file_storage.file.streamed_file_fetcher'
             - '@pim_api.repository.product'
+            - '@pim_catalog.repository.product_model'
```
Comment (momoss) [inline src/Pim/Bundle/ApiBundle/Resources/config/controllers.yml]: Maybe add your repo at the end to limit BC breaks

File: src/Pim/Component/Catalog/Validator/Constraints/OnlyExpectedAttributesValidator.php [hunk 3]
```diff
@@ -61,7 +61,7 @@ public function validate($entity, Constraint $constraint): void
                 $this->context->buildViolation(
                     OnlyExpectedAttributes::ATTRIBUTE_UNEXPECTED, [
                     '%attribute%' => $modelAttribute->getCode()
-                ])->addViolation();
+                ])->atPath('values')->addViolation();
```
Comment (momoss) [inline src/Pim/Component/Catalog/Validator/Constraints/OnlyExpectedAttributesValidator.php]: Why did you need to specify the path?

Comment (willy-ahva) [reply src/Pim/Component/Catalog/Validator/Constraints/OnlyExpectedAttributesValidator.php]: For error message

File: src/Pim/Bundle/ApiBundle/Resources/config/controllers.yml (hunk 2 above)
Comment (willy-ahva) [reply src/Pim/Bundle/ApiBundle/Resources/config/controllers.yml]: Seen with nidup, BC don't matter here

File: CHANGELOG-2.0.md (hunk 1 above)
Comment (willy-ahva) [reply CHANGELOG-2.0.md]: nope

File: src/Pim/Bundle/ApiBundle/Resources/config/controllers.yml (hunk 2 above)
Comment (momoss) [reply src/Pim/Bundle/ApiBundle/Resources/config/controllers.yml]: Need to overridde controller declaration in EE

**Comment by momoss** [comment]
[CI broken](https://ci.akeneo.com/blue/organizations/jenkins/akeneo%2Fpim-community-dev/detail/PR-7419/5/tests)

File: src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php [hunk 4]
```diff
@@ -370,4 +485,28 @@ protected function getDecodedContent($content)
...
     }
+
+    /**
+     * @param string $content
+     *
+     * @throws HttpException
+     *
+     * @return array
+     */
+    protected function getProductModelDecodedContent($content)
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php]: return array

File: src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php [hunk 5]
```diff
@@ -316,6 +387,50 @@ protected function linkFileToProduct(FileInfoInterface $fileInfo, ProductInterfa
...
+    /**
+     * @param FileInfoInterface     $fileInfo
+     * @param ProductModelInterface $productModel
+     * @param array                 $productModelInfos
+     *
+     * @throws HttpException
+     *
+     * @return ProductModelInterface
+     */
+    protected function linkFileToProductModel(
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php]: return ProductModelInterface

File: src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php [hunk 6]
```diff
@@ -354,7 +469,7 @@ protected function storeFile(FileBag $files)
      *
      * @return array
      */
-    protected function getDecodedContent($content)
+    protected function getProductDecodedContent($content)
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php]: return array

File: tests/catalog/technical/families.csv [hunk 7]
```diff
@@ -1,5 +1,5 @@
...
+familyA;Une famille A;A family A;a_date,a_file,a_localizable_image,a_localizable_scopable_image,a_localized_and_scopable_text_area,a_metric,a_multi_select,a_number_float,a_number_float_negative,a_number_integer,a_price,a_ref_data_multi_select,a_ref_data_simple_select,a_scopable_price,a_simple_select,a_text,a_text_area,a_yes_no,an_image,sku;sku;a_date,a_file,a_localizable_image,a_localized_and_scopable_text_area,a_metric,a_multi_select,a_number_float,a_number_float_negative,a_number_integer,a_price,a_ref_data_multi_select,a_ref_data_simple_select,a_scopable_price,a_simple_select,a_text,a_text_area,a_yes_no,an_image,sku;a_date,a_file,a_localizable_image,a_localized_and_scopable_text_area,a_metric,a_multi_select,a_number_float,a_number_float_negative,a_number_integer,a_price,a_ref_data_multi_select,a_ref_data_simple_select,a_scopable_price,a_simple_select,a_text,a_text_area,a_yes_no,an_image,sku;an_image
```
Comment (ahocquard) [inline tests/catalog/technical/families.csv]: Create a family just for your use case. Don't modify the whole family just for it. We have already too much things in the catalog.

File: src/Pim/Bundle/ApiBundle/tests/integration/Controller/Media/CreateProductModelMediaFileIntegration.php [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+                        'a_simple_select' => [['data' => 'optionA', 'locale' => null, 'scope' => null]]
+                    ]
+                ]
+            );
+        $this->assertCount(0, $this->getFromTestContainer('pim_catalog.validator.product_model')->validate($productModel));
+        $this->getFromTestContainer('pim_catalog.saver.product_model')->save($productModel);
+
+        $this->getFromTestContainer('akeneo_elasticsearch.client.product_model')->refreshIndex();
+
+        $this->getFromTestContainer('akeneo_storage_utils.doctrine.object_detacher')->detach($familyVariant);
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/tests/integration/Controller/Media/CreateProductModelMediaFileIntegration.php]: clear doctrine to have a completely clean state

File: src/Pim/Bundle/ApiBundle/tests/integration/Controller/Media/CreateProductModelMediaFileIntegration.php [hunk 9]
```diff
@@ -0,0 +1,630 @@
...
+                    'a_scopable_price',
+                    'a_simple_select',
+                    'a_text',
+                    'a_text_area',
+                    'a_yes_no',
+                    'an_image',
+                    'a_localizable_scopable_image'
+            ]
+        ]);
+        $this->getFromTestContainer('pim_catalog.saver.family')->save($familyA);
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/tests/integration/Controller/Media/CreateProductModelMediaFileIntegration.php]: validate it before please

File: src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php [hunk 10]
```diff
@@ -119,7 +132,10 @@ public function __construct(
         FileStorerInterface $fileStorer,
         RemoverInterface $remover,
         RouterInterface $router,
-        array $apiConfiguration
+        array $apiConfiguration,
```
Comment (momoss) [inline src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php]: Always put non Object Params at the end

**Comment by momoss** [comment]
Don't forget to update the PR description 😄
